File: lua/pov.lua
local function to_nodes(tree, nodes)
    assert(#tree > 0)
    local parent = tree[1]
    nodes[parent] = nodes[parent] or {}

    for _, child in ipairs(tree[2] or {}) do
        table.insert(nodes[parent], child[1])
        to_nodes(child, nodes)
        table.insert(nodes[child[1]], parent)
    end
end

local function build_tree(nodes, curr, parent)
    local t = { curr }
    for _, node in ipairs(nodes[curr]) do
        if node ~= parent then
            t[2] = t[2] or {}
            table.insert(t[2], build_tree(nodes, node, curr))
        end
    end
    return t
end

local function pov_from(node_name)
    return {
        of = function(tree)
            local nodes = {}
            to_nodes(tree, nodes)
            assert(nodes[node_name])
            return build_tree(nodes, node_name, nil)
        end,
    }
end

local function dfs(nodes, curr, target, visited)
    if curr == target then
        return { target }
    end

    if visited[curr] ~= nil then
        return nil
    end
    visited[curr] = 1
    for _, each in ipairs(nodes[curr]) do
        local p = dfs(nodes, each, target, visited)
        if p ~= nil then
            table.insert(p, 1, curr)
            return p
        end
    end
    return nil
end

local function path_from(source)
    return {
        to = function(target)
            return {
                of = function(tree)
                    local nodes = {}
                    to_nodes(tree, nodes)
                    assert(nodes[source] and nodes[target])
                    return dfs(nodes, source, target, {})
                end,
            }
        end,
    }
end

return { pov_from = pov_from, path_from = path_from }

File: python/pov.py
from json import dumps


class Tree:
    def __init__(self, label, children=None):
        self.label = label
        self.children = children if children is not None else []

    def __dict__(self):
        return {self.label: [c.__dict__() for c in sorted(self.children)]}

    def __str__(self, indent=None):
        return dumps(self.__dict__(), indent=indent)

    def __lt__(self, other):
        return self.label < other.label

    def __eq__(self, other):
        return self.__dict__() == other.__dict__()

    def to_nodes(self, nodes):
        for child in self.children:
            nodes[self.label].append(child.label)
            nodes[child.label] = [self.label]
            child.to_nodes(nodes)

    @staticmethod
    def from_nodes(nodes, curr, parent):
        t = Tree(curr)
        for each in nodes[curr]:
            if each != parent:
                t.children.append(Tree.from_nodes(nodes, each, curr))
        return t

    def from_pov(self, from_node):
        nodes = {self.label: []}
        self.to_nodes(nodes)
        if from_node not in nodes:
            raise ValueError("Tree could not be reoriented")
        return Tree.from_nodes(nodes, from_node, None)

    def find_path(self, to_node, visited):
        print(self.label)
        if self.label == to_node:
            return [to_node]
        if self.label in visited:
            return None
        visited.add(self.label)
        for child in self.children:
            path = child.find_path(to_node, visited)
            if path is None:
                continue
            path.append(self.label)
            return path
        return None

    def path_to(self, from_node, to_node):
        path = self.from_pov(from_node).find_path(to_node, set())
        if path is None:
            raise ValueError("No path found")
        return path[::-1]
